File: application/backend/business/rules-processor-v1/src/main/java/com/demo/poc/commons/core/interceptor/error/ResponseErrorSelector.java
package com.demo.poc.commons.core.interceptor.error;

import com.demo.poc.commons.core.errors.dto.ErrorDTO;
import com.demo.poc.commons.core.errors.dto.ErrorType;
import com.demo.poc.commons.core.errors.exceptions.GenericException;
import com.demo.poc.commons.core.properties.ConfigurationBaseProperties;
import com.demo.poc.commons.core.properties.ProjectType;
import lombok.AccessLevel;
import lombok.NoArgsConstructor;
import org.springframework.web.bind.MethodArgumentNotValidException;

import java.util.Optional;
import java.util.stream.Collectors;

import static com.demo.poc.commons.core.errors.dto.ErrorDTO.CODE_DEFAULT;

@NoArgsConstructor(access = AccessLevel.PRIVATE)
public class ResponseErrorSelector {

  public static <T extends Throwable> ErrorDTO toErrorDTO(ConfigurationBaseProperties properties, T exception) {
    ErrorDTO error = extractError(exception);
    String selectedCode = selectCustomCode(properties, error);
    error.setCode(selectedCode);

    String selectedMessage = selectMessage(properties, error);
    error.setMessage(selectedMessage);

    return error;
  }

  private static String selectCustomCode(ConfigurationBaseProperties properties, ErrorDTO error) {
    ProjectType projectType = selectProjectType(properties);

    return ProjectType.BFF.equals(projectType)
        ? CODE_DEFAULT
        : Optional.ofNullable(error)
        .map(ErrorDTO::getCode)
        .orElseGet(() -> CODE_DEFAULT);
  }

  private static String selectMessage(ConfigurationBaseProperties properties, ErrorDTO error) {
    String defaultMessage = ErrorDTO.getDefaultError(properties).getMessage();
    ProjectType projectType = selectProjectType(properties);

    if (ProjectType.BFF.equals(projectType))
      return defaultMessage;

    return Optional.ofNullable(ErrorDTO.getMatchMessage(properties, error.getCode()))
        .orElseGet(() -> Optional.ofNullable(error.getMessage()).orElse(defaultMessage));
  }

  private static ProjectType selectProjectType(ConfigurationBaseProperties properties) {
    return Optional.ofNullable(properties.getProjectType()).orElseGet(() -> ProjectType.MS);
  }

  private static <T extends Throwable> ErrorDTO extractError(T exception) {
    ErrorDTO error = null;

    if (exception instanceof GenericException genericException)
      error = genericException.getErrorDetail();

    if (exception instanceof MethodArgumentNotValidException methodArgumentNotValidExceptionD) {
      error = ErrorDTO.builder()
          .type(ErrorType.BUSINESS)
          .message(methodArgumentNotValidExceptionD.getBindingResult()
              .getFieldErrors()
              .stream()
              .map(fieldError -> fieldError.getField() + ": " + fieldError.getDefaultMessage())
              .collect(Collectors.joining(";")))
          .code("InvalidInputField")
          .build();
    }

    return error;
  }
}

File: backend/business/product-v1/commons/errors/selector/response_error_selector.go
package selector

import (
	"com.demo.poc/commons/constants"
	errorDto "com.demo.poc/commons/errors/dto"
	properties "com.demo.poc/commons/properties"
)

type ResponseErrorSelector struct {
	properties *properties.ApplicationProperties
}

func NewResponseErrorSelector(properties *properties.ApplicationProperties) *ResponseErrorSelector {
	return &ResponseErrorSelector{properties: properties}
}

func (responseErrorSelector *ResponseErrorSelector) ToErrorDto(err error) errorDto.ErrorDto {
	baseError := extractError(err)
	projectType := responseErrorSelector.selectProjectType()
	baseError.Code = selectCustomCode(baseError.Code, projectType)
	baseError.Message = responseErrorSelector.selectMessage(baseError.Message, baseError.Code, projectType)
	return baseError
}

func extractError(err error) errorDto.ErrorDto {
	if genericError, isGenericError := err.(errorDto.GenericError); isGenericError {
		recoveredError := errorDto.ErrorDto{
			Origin: string(genericError.Origin),
			Code:   genericError.Code,
		}
		if message := genericError.Message; message != constants.EMPTY {
			recoveredError.Message = message
		}
		return recoveredError
	}
	return errorDto.ErrorDto{Origin: string(errorDto.ERROR_ORIGIN_OWN)}
}

func (responseErrorSelector *ResponseErrorSelector) selectProjectType() properties.ProjectType {
	if projectType := properties.ProjectType(responseErrorSelector.properties.ProjectType); projectType != constants.EMPTY {
		return projectType
	}
	return properties.PROJECT_TYPE_MS
}

func selectCustomCode(code string, projectType properties.ProjectType) string {
	if projectType == properties.PROJECT_TYPE_BFF {
		return errorDto.CODE_DEFAULT
	}
	if code == constants.EMPTY {
		return errorDto.CODE_DEFAULT
	}
	return code
}

func (responseErrorSelector *ResponseErrorSelector) selectMessage(originalMessage, code string, projectType properties.ProjectType) string {
	defaultMessage := responseErrorSelector.properties.ErrorMessages[errorDto.CODE_DEFAULT]

	if projectType == properties.PROJECT_TYPE_BFF {
		return defaultMessage
	}
	if message, existsMessage := responseErrorSelector.properties.ErrorMessages[code]; existsMessage {
		return message
	}
	if originalMessage != constants.EMPTY {
		return originalMessage
	}
	return defaultMessage
}
